tracer/src/Datadog.Trace/AppSec/RateLimiterTimer.cs

```diff
@@ -0,0 +1,41 @@
+// <copyright file="RateLimiterTimer.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Threading;
+
+namespace Datadog.Trace.AppSec
+{
+    internal class RateLimiterTimer : IDisposable
+    {
+        private readonly Timer _timer;
+
+        private int _rateLimiterCounter;
+        private int _exceededTraces;
+
+        public RateLimiterTimer(ref int rateLimiterCounter, ref int exceededTraces)
```

Comment:
It doesn't look like these need to be `ref` parameters. Was there a reason for this?

Reply:
Oh, I think you're trying to modify the field `Security.rateLimiterCounter` from here. But the `_rateLimiterCounter` and `_exceededTraces` will be copies, not references/pointers.

I think a better way to do this would be to expose public members in `RateLimiterTimer`:
```csharp
int RateLimiterCounter { get; }
int ExceededTraces { get; }
void IncrementRateLimiterCounter();
void IncrementExceededTraces();
```

Or maybe something different? Hard for me to think about it without writing the code 😅

Reply:
I agree with Lucas that I think there might be a better way to express this logic. One suggestion at the top of my head would be to introduce a new method:

```
// exceededTracesCount is populated only when returning false
bool Allowed(out int exceededTracesCount);
```

This method could replace the `if (rateLimiterCounter < _settings.TraceRateLimit)` check and all of the other members (including `_settings.TraceRateLimit`) could be stored internally in the rate limiter. That would be one way to try isolate the fields to avoid sharing them by ref

Reply:
thanks, fixed in fixed in 8a1b7d7cfb8567399adc367e4bb2c284b1a35ea7 and 0981288cf418e39161b6f681fc457ad8209886c2